payments-core/src/main/java/com/stripe/android/googlepaylauncher/GooglePayPaymentMethodLauncherViewModel.kt

```diff
@@ -0,0 +1,169 @@
+import kotlinx.coroutines.flow.first
+import org.json.JSONObject
+import kotlin.coroutines.CoroutineContext
+
+internal class GooglePayPaymentMethodLauncherViewModel(
+    private val paymentsClient: PaymentsClient,
+    private val requestOptions: ApiRequest.Options,
+    private val args: GooglePayPaymentMethodLauncherContract.Args,
+    private val stripeRepository: StripeRepository,
+    private val googlePayJsonFactory: GooglePayJsonFactory,
+    private val googlePayRepository: GooglePayRepository
+) : ViewModel() {
+    var hasLaunched: Boolean = false
+
+    private val _googleResult = MutableLiveData<GooglePayPaymentMethodLauncher.Result>()
+    internal val googlePayResult = _googleResult.distinctUntilChanged()
+
+    fun updateResult(result: GooglePayPaymentMethodLauncher.Result) {
+        _googleResult.value = result
+    }
+
+    @VisibleForTesting
+    suspend fun isReadyToPay(): Boolean {
+        return googlePayRepository.isReady().first()
+    }
+
+    @VisibleForTesting
+    fun createPaymentDataRequest(): JSONObject {
+        return googlePayJsonFactory.createPaymentDataRequest(
+            transactionInfo = createTransactionInfo(args),
+            merchantInfo = GooglePayJsonFactory.MerchantInfo(
+                merchantName = args.config.merchantName
+            ),
+            billingAddressParameters = GooglePayJsonFactory.BillingAddressParameters(
+                isRequired = true,
+                format = GooglePayJsonFactory.BillingAddressParameters.Format.Min,
+                isPhoneNumberRequired = false
+            ),
+            isEmailRequired = args.config.isEmailRequired
+        )
+    }
+
+    @VisibleForTesting
+    internal fun createTransactionInfo(
+        args: GooglePayPaymentMethodLauncherContract.Args
+    ): GooglePayJsonFactory.TransactionInfo {
+        return GooglePayJsonFactory.TransactionInfo(
+            currencyCode = args.currencyCode,
+            totalPriceStatus = when (args.hasAmount) {
+                true -> GooglePayJsonFactory.TransactionInfo.TotalPriceStatus.Final
+
+                false -> GooglePayJsonFactory.TransactionInfo.TotalPriceStatus.NotCurrentlyKnown
+            },
+            countryCode = args.config.merchantCountryCode,
+            totalPrice = args.amount,
+            checkoutOption = GooglePayJsonFactory.TransactionInfo.CheckoutOption.Default
+        )
+    }
+
+    suspend fun createLoadPaymentDataTask(): Task<PaymentData> {
+        check(isReadyToPay()) {
+            "Google Pay is unavailable."
+        }
+        return paymentsClient.loadPaymentData(
+            PaymentDataRequest.fromJson(createPaymentDataRequest().toString())
+        )
+    }
+
+    suspend fun createPaymentMethod(
+        paymentData: PaymentData
+    ): GooglePayPaymentMethodLauncher.Result {
+        val paymentDataJson = JSONObject(paymentData.toJson())
+
+        val params = PaymentMethodCreateParams.createFromGooglePay(paymentDataJson)
```

Comment:
I wonder if createFromGooglePay should take a GooglePay object?  Seems like anything that calls the function is going to need to do the ```val paymentDataJson = JSONObject(paymentData.toJson())```

Reply:
This would add a dependency on Google Pay in the core API, which we probably don't want to do.

Reply:
Do you see the PaymentMethodCreateParams.createFromGooglePay moving to a GooglePay only module?

Reply:
It's still convenient for users rolling their own Google Pay integration.

Reply:
I suppose users could still access it if it was in a different module.

Reply:
Yeah maybe something we can look at in the future.